src/fifo.c

```diff
@@ -0,0 +1,299 @@
+         *   |                  (61 bits)                     | (3 bits) |
+         *   +-----------------------------------------------------------+
+         *     * The next pointer is valid as-is
+         *     * "lastIdx" is implied to be 6 in all middle blocks.
+         *     * "firstIdx" is implied to be 0 in all middle blocks.
+         *     * NOTE: In middle blocks, the index bits(0) are really still the firstIdx value.
+         *             When Fifo's are joined, the O(1) operation may result in a partially
+         *             full middle block.  In this case, the items are "right-justified" and
+         *             firstIdx indicates where the items start.
+         *
+         * The last (or only) block in the chain contains only the lastIndex, the pointer is unused.
+         *   +-----------------------------------------------------------+
+         *   |                      0                         | lastIdx  |
+         *   |                  (61 bits)                     | (3 bits) |
+         *   +-----------------------------------------------------------+
+         *     * The next pointer is unused and guaranteed NULL.
+         *     * "lastIdx" represents the last filled index (0..6).
+         *     * "firstIdx" is implied to be zero on the last (or only) block.
+         */
+        uintptr_t last_or_first_idx;
+        struct FifoBlock *next;
+    } u;
+} FifoBlock;
+
+struct Fifo {
+    long length;        // Total number of items in queue
+    FifoBlock *first;
+    FifoBlock *last;
+};
+
+
+// Create a new FIFO queue.
+Fifo *fifoCreate() {
+    Fifo *q = zmalloc(sizeof(Fifo));
+    q->length = 0;
+    q->first = q->last = NULL;
+    return q;
+}
+
+
+// Push an item onto the end of the queue.
+void fifoPush(Fifo *q, void *ptr) {
+    if (q->first == NULL) {
+        // Queue was empty - create block
+        assert(q->last == NULL && q->length == 0);
+        q->last = q->first = zmalloc(sizeof(FifoBlock));
+        q->last->u.last_or_first_idx = 0;   // Item 0 is the last item in this block
+        q->last->items[0] = ptr;
+    } else {
+        int lastIdx = q->last->u.last_or_first_idx; // pointer portion is 0 on last (or only) block
+        assert(lastIdx < ITEMS_PER_BLOCK);
+
+        if (lastIdx < ITEMS_PER_BLOCK - 1) {
+            // If the last block has space, just add the item
+            q->last->items[lastIdx + 1] = ptr;
+            q->last->u.last_or_first_idx++;
+        } else {
+            // Otherwise, last block is full - add a new block
+            FifoBlock *newblock = zmalloc(sizeof(FifoBlock));
+            newblock->u.last_or_first_idx = 0;
+            newblock->items[0] = ptr;
+            q->last->u.next = newblock;     // overwrites the index, setting it to 0
+            q->last = newblock;
+        }
+    }
+
+    q->length++;
+}
+
+
+// Look at the first item in the queue (without removing it).
+// NOTE: asserts if the queue is empty.
```

Comment:
I'm not sure I fully buy the utility of asserting on the queue being non-empty, most implementations would return NULL instead. You implement this paradigm by getting the length first, and then deciding to call peek.

Reply:
Since the value is user-defined, 0/NULL might be a valid value.  You could change the API to look something like:
```
bool fifoPeek(Fifo *q, void **returned_value);
```
This definition is uglier, and the caller still has to check the bool (which is not much different than checking the length).  [And if we still can't use `bool`, it's even uglier with an `int`!]

Reply:
I was just assuming you could require end users to wrap objects that might be NULL or return a value that embeds the pointer, but I like the proposed API too. Having an API that "might" sometimes assert is a very difficult API to understand and might introduce non-obvious bugs. Forcing a user to also check for the length puts a lot of complexity onto the user to remember.